nexus/db-queries/src/db/datastore/update.rs

```diff
@@ -195,6 +200,220 @@ impl DataStore {
+                    // If the target release generation has changed, bail out.
+                    // This means someone has changed the target release, which
+                    // means they may have set it to the repo we're trying to
+                    // prune.  (This check could be more fine-grained, but this
+                    // is adequate for now)
+                    let target_release_generation_now = {
+                        use nexus_db_schema::schema::target_release::dsl;
+                        dsl::target_release
+                            .select(TargetRelease::as_select())
+                            .order_by(dsl::generation.desc())
+                            .limit(1)
+                            .first_async(&txn)
+                            .await
+                            .map_err(|e| {
+                                public_error_from_diesel(
+                                    e,
+                                    ErrorHandler::Server,
+                                )
+                            })
+                            .internal_context(
+                                "fetching latest target_release generation",
+                            )
+                            .map_err(|e| {
+                                error.bail(TransactionError::CustomError(e))
+                            })?
+                            .generation
+                            .0
+                    };
+                    if target_release_generation_now
+                        != recent_releases.target_release_generation
+                    {
+                        return Err(error.bail(TransactionError::CustomError(
+                            Error::conflict(format!(
+                                "bailing out to avoid risk of marking current \
+                                 target release pruned: target release has \
+                                 changed since check (currently {}, was {})",
+                                target_release_generation_now,
+                                recent_releases.target_release_generation
+                            )),
+                        )));
+                    }
+
+                    // If the TUF repo generation has changed, bail out.
+                    // Someone else is adding or pruning repos.  Force the
+                    // caller to re-evaluate.  This is probably more
+                    // conservative than necessary, but ensures that two Nexus
+                    // instances don't concurrently decide to prune a lot more
+                    // than either of them would on their own because they chose
+                    // different repos to keep.
+                    let tuf_generation_now = get_generation(&txn)
+                        .await
+                        .map_err(|e| {
+                            public_error_from_diesel(e, ErrorHandler::Server)
+                        })
```

Comment:
Same here, this would convert any retryable errors into an externally-facing error, like a 500.